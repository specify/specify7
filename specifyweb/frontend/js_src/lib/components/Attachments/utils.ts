import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { genericTables, getTable } from '../DataModel/tables';
import { hasTablePermission } from '../Permissions/helpers';

export const attachmentRelatedTables = f.store(() =>
  Object.keys(genericTables).filter((tableName) =>
    tableName.endsWith('Attachment')
  )
);

export const allTablesWithAttachments = f.store(() =>
  filterArray(
    attachmentRelatedTables().map((tableName) =>
      getTable(tableName.slice(0, -1 * 'Attachment'.length))
    )
  )
);

/** Exclude tables without read access */
export const tablesWithAttachments = f.store(() =>
  allTablesWithAttachments().filter((model) =>
    hasTablePermission(model.name, 'read')
  )
);

/**
 * Gets the relationship to a Table's attachment table.
 * e.g., the collectionObjectAttachments relationship from CollectionObject to CollectionObjectAttachments
 * In most cases the relationship is called tableNameAttachments, although fetching the relationship
 * following that schema will not always be consitent as the following models relationship to their
 * attachment table is simply called `attachments`:
 * - DNASequence
 * - FieldNotebook
 * - FieldNotebookPage
 * - FieldNotebookPageSet
 */
export const getAttachmentRelationship = (
  table: SpecifyTable
): Relationship | undefined => {
  if (table.name === 'Attachment') return undefined;
  const commonRelationship = table.field[
    `${table.name[0].toLowerCase()}${table.name.slice(1)}Attachments`
  ] as Relationship | undefined;

  return commonRelationship === undefined
    ? table.relationships.find((relationship) => {
        const relatedModel = relationship.relatedTable.name;
        return (
          relatedModel !== 'Attachment' && relatedModel.endsWith('Attachment')
        );
      })
    : commonRelationship;
};
