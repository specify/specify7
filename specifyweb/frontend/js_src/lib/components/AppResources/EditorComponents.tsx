import { json } from '@codemirror/lang-json';
import { xml } from '@codemirror/lang-xml';
import { indentUnit, StreamLanguage } from '@codemirror/language';
import { properties } from '@codemirror/legacy-modes/mode/properties';
import type { Diagnostic } from '@codemirror/lint';
import { lintGutter, openLintPanel } from '@codemirror/lint';
import { search } from '@codemirror/search';
import type { Extension } from '@codemirror/state';
import { EditorState } from '@codemirror/state';
import { EditorView } from 'codemirror';
import React from 'react';
import type { LocalizedString } from 'typesafe-i18n';

import { useBooleanState } from '../../hooks/useBooleanState';
import { commonText } from '../../localization/common';
import { notificationsText } from '../../localization/notifications';
import { resourcesText } from '../../localization/resources';
import type { RA } from '../../utils/types';
import { filterArray } from '../../utils/types';
import { Button } from '../Atoms/Button';
import { DataEntry } from '../Atoms/DataEntry';
import { LoadingContext } from '../Core/Contexts';
import { getField } from '../DataModel/helpers';
import type { SerializedResource } from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import { getFieldBlockerKey, useSaveBlockers } from '../DataModel/saveBlockers';
import type { SpAppResource, SpViewSetObj } from '../DataModel/types';
import { DeleteButton } from '../Forms/DeleteButton';
import { Dialog } from '../Molecules/Dialog';
import { downloadFile, FilePicker, fileToText } from '../Molecules/FilePicker';
import { hasToolPermission } from '../Permissions/helpers';
import { userPreferences } from '../Preferences/userPreferences';
import type { BaseSpec } from '../Syncer';
import type { SimpleXmlNode } from '../Syncer/xmlToJson';
import { jsonLinter, xmlLinter } from './codeMirrorLinters';
import type { getResourceType } from './filtersHelpers';
import { getAppResourceExtension } from './hooks';
import { appResourceSubTypes, appResourceTypes } from './types';

export const appResourceIcon = (
  type: ReturnType<typeof getResourceType>
): JSX.Element => (
  <span
    /*
     * FEATURE: include this title in aria-describedby on the button itself
     *   so that screen readers announce the type after the name
     */
    aria-hidden
    title={
      type === 'viewSet'
        ? resourcesText.formDefinitions()
        : appResourceSubTypes[type].label
    }
  >
    {type === 'viewSet'
      ? appResourceTypes.viewSets.icon
      : appResourceSubTypes[type].icon}
  </span>
);

export function AppResourceEditButton({
  title,
  children,
  appResource,
  onDeleted: handleDeleted,
}: {
  readonly title: LocalizedString;
  readonly children: JSX.Element;
  readonly appResource: SpecifyResource<SpAppResource>;
  readonly onDeleted: (() => void) | undefined;
}): JSX.Element {
  const [isEditingMeta, handleEditingMeta, handleEditedMeta] =
    useBooleanState();
  return (
    <>
      <DataEntry.Edit onClick={handleEditingMeta} />
      {isEditingMeta && (
        <Dialog
          buttons={
            <>
              {!appResource.isNew() &&
              hasToolPermission('resources', 'delete') &&
              typeof handleDeleted === 'function' ? (
                <DeleteButton
                  resource={appResource}
                  onDeleted={handleDeleted}
                />
              ) : undefined}
              <span className="-ml-2 flex-1" />
              <Button.DialogClose>{commonText.close()}</Button.DialogClose>
            </>
          }
          dimensionsKey="AppResourceEdit"
          header={title}
          onClose={handleEditedMeta}
        >
          {children}
        </Dialog>
      )}
    </>
  );
}

export function AppResourceLoad({
  onLoaded: handleLoaded,
}: {
  readonly onLoaded: (data: string, mimeType: string) => void;
}): JSX.Element {
  const [isOpen, handleOpen, handleClose] = useBooleanState();
  const loading = React.useContext(LoadingContext);
  return (
    <>
      <Button.Success className="whitespace-nowrap" onClick={handleOpen}>
        {resourcesText.loadFile()}
      </Button.Success>
      {isOpen && (
        <Dialog
          buttons={commonText.cancel()}
          header={resourcesText.loadFile()}
          onClose={handleClose}
        >
          <FilePicker
            acceptedFormats={undefined}
            onFileSelected={(file): void =>
              loading(
                fileToText(file)
                  .then((data) => handleLoaded(data, file.type))
                  .finally(handleClose)
              )
            }
          />
        </Dialog>
      )}
    </>
  );
}

export function AppResourceDownload({
  resource,
  data,
}: {
  readonly resource: SerializedResource<SpAppResource | SpViewSetObj>;
  readonly data: string;
}): JSX.Element {
  const loading = React.useContext(LoadingContext);
  return (
    <Button.Success
      className="whitespace-nowrap"
      disabled={data.length === 0}
      onClick={(): void =>
        loading(
          downloadFile(
            `${resource.name}.${getAppResourceExtension(resource)}`,
            data
          )
        )
      }
    >
      {notificationsText.download()}
    </Button.Success>
  );
}

export function useIndent(): string {
  const [indentSize] = userPreferences.use(
    'appResources',
    'behavior',
    'indentSize'
  );
  const [indentWithTab] = userPreferences.use(
    'appResources',
    'behavior',
    'indentWithTab'
  );
  return indentWithTab ? '\t' : ' '.repeat(indentSize);
}

/**
 * Use useIndent() instead whenever possible
 */
export function getIndent(): string {
  const indentSize = userPreferences.get(
    'appResources',
    'behavior',
    'indentSize'
  );
  const indentWithTab = userPreferences.get(
    'appResources',
    'behavior',
    'indentWithTab'
  );
  return indentWithTab ? '\t' : ' '.repeat(indentSize);
}

export function useCodeMirrorExtensions(
  resource: SerializedResource<SpAppResource | SpViewSetObj>,
  appResource: SpecifyResource<SpAppResource | SpViewSetObj>,
  xmlSpec: (() => BaseSpec<SimpleXmlNode>) | undefined
): RA<Extension> {
  const [lineWrap] = userPreferences.use(
    'appResources',
    'behavior',
    'lineWrap'
  );
  const [indentSize] = userPreferences.use(
    'appResources',
    'behavior',
    'indentSize'
  );
  const indentCharacter = useIndent();

  const mode = React.useMemo(
    () => getAppResourceExtension(resource),
    [resource]
  );
  const field = React.useMemo(
    () => getField(appResource.specifyTable, 'spAppResourceDatas'),
    [appResource.specifyTable]
  );

  const [extensions, setExtensions] = React.useState<RA<Extension>>([]);
  const [_blockers, setBlockers] = useSaveBlockers(appResource, field);

  React.useEffect(() => {
    let isFirstLint = true;
    function handleLinted(results: RA<Diagnostic>, view: EditorView): void {
      if (isFirstLint && results.length > 0) {
        isFirstLint = false;
        setTimeout(() => {
          const currentFocus = document.activeElement as HTMLElement | null;
          openLintPanel(view);
          currentFocus?.focus();
        }, 0);
      }
      setBlockers(
        filterArray(
          results.map(({ message, severity }) =>
            severity === 'error' ? message : undefined
          )
        ),
        getFieldBlockerKey(field, 'appResourceError')
      );
    }

    const language =
      mode === 'json'
        ? [json(), jsonLinter(handleLinted)]
        : mode === 'properties'
          ? [StreamLanguage.define(properties)]
          : mode === 'jrxml' || mode === 'xml'
            ? [xml(), xmlLinter(xmlSpec?.())(handleLinted)]
            : [];
    setExtensions([
      ...language,
      ...(lineWrap ? [EditorView.lineWrapping] : []),
      indentUnit.of(indentCharacter),
      EditorState.tabSize.of(indentSize),
      lintGutter(),
      search(),
    ]);

    return (): void =>
      setBlockers([], getFieldBlockerKey(field, 'appResourceError'));
  }, [
    appResource,
    mode,
    lineWrap,
    indentCharacter,
    indentSize,
    xmlSpec,
    setBlockers,
  ]);

  return extensions;
}
