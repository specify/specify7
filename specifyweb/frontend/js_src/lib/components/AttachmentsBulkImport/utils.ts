import type { State } from 'typesafe-reducer';

import { attachmentsText } from '../../localization/attachments';
import { commonText } from '../../localization/common';
import { formsText } from '../../localization/forms';
import { wbText } from '../../localization/workbench';
import type { AjaxResponseObject } from '../../utils/ajax';
import { ajax } from '../../utils/ajax';
import { Http } from '../../utils/ajax/definitions';
import { f } from '../../utils/functools';
import type { RA, RR } from '../../utils/types';
import { defined, filterArray } from '../../utils/types';
import {
  insertItem,
  keysToLowerCase,
  mappedFind,
  replaceItem,
  stripFileExtension,
} from '../../utils/utils';
import { addMissingFields } from '../DataModel/addMissingFields';
import type {
  SerializedRecord,
  SerializedResource,
} from '../DataModel/helperTypes';
import type { SpecifyResource } from '../DataModel/legacyTypes';
import {
  deserializeResource,
  serializeResource,
} from '../DataModel/serializers';
import { strictGetTable, tables } from '../DataModel/tables';
import type { SpQuery, Tables } from '../DataModel/types';
import type { CollectionObject } from '../DataModel/types';
import type { UiFormatter } from '../FieldFormatters';
import { formatterTypeMapper } from '../FieldFormatters';
import { queryFieldFilters } from '../QueryBuilder/FieldFilter';
import { makeQueryField } from '../QueryBuilder/fromTree';
import type { QueryFieldWithPath } from '../Statistics/types';
import type { AttachmentUploadSpec } from './Import';
import { staticAttachmentImportPaths } from './importPaths';
import type { AttachmentStatus, PartialUploadableFileSpec } from './types';

export type ResolvedAttachmentRecord =
  | State<
      'invalid' | 'valid',
      { readonly reason: keyof typeof keyLocalizationMapAttachment }
    >
  | State<'matched', { readonly id: number }>;

const resolveAttachmentMatch = (
  matchedId: RA<number>,
  disambiguated: number | undefined
): ResolvedAttachmentRecord =>
  matchedId.length === 0
    ? { type: 'invalid', reason: 'noMatch' }
    : matchedId.length > 1 && disambiguated === undefined
    ? {
        type: 'invalid',
        reason: 'multipleMatches',
      }
    : {
        type: 'matched',
        id: disambiguated ?? matchedId[0],
      };

export function resolveAttachmentRecord(
  matchedId: RA<number> | undefined,
  disambiguated: number | undefined,
  parsedName: string | undefined
): ResolvedAttachmentRecord {
  if (parsedName === undefined)
    return { type: 'invalid', reason: 'incorrectFormatter' };
  if (matchedId === undefined)
    return { type: 'valid', reason: 'correctlyFormatted' };
  return resolveAttachmentMatch(matchedId, disambiguated);
}

export const canDeleteAttachment = (
  uploadSpec: PartialUploadableFileSpec
): boolean =>
  uploadSpec.attachmentId !== undefined &&
  uploadSpec.matchedId !== undefined &&
  resolveAttachmentMatch(uploadSpec.matchedId, uploadSpec.disambiguated)
    .type === 'matched';

type InQueryField = RA<{
  readonly field: QueryFieldWithPath;
  readonly lookUp?: RA<number | string | undefined>;
}>;

function generateInQueryResource(
  baseTable: keyof Tables,
  queryName: string,
  inQueryFields: InQueryField
): SpecifyResource<SpQuery> {
  const queryFields = inQueryFields.map((inQueryField, index) => {
    const rawField =
      inQueryField.lookUp === undefined
        ? inQueryField.field
        : {
            isDisplay: true,
            operStart: queryFieldFilters.in.id,
            // Just unique values are necessary. Also decreases the number of values to send to backend
            startValue: f.unique(filterArray(inQueryField.lookUp)).join(','),
            ...inQueryField.field,
          };
    const { path, ...field } = rawField;
    return serializeResource(
      makeQueryField(baseTable, rawField.path, { ...field, position: index })
    );
  });

  return deserializeResource(
    addMissingFields('SpQuery', {
      name: queryName,
      contextName: baseTable,
      contextTableId: tables[baseTable].tableId,
      countOnly: false,
      fields: queryFields,
    })
  );
}

export async function validateAttachmentFiles(
  uploadableFiles: RA<PartialUploadableFileSpec>,
  uploadSpec: AttachmentUploadSpec,
  keepDisambiguation: boolean = false
): Promise<RA<PartialUploadableFileSpec>> {
  const { baseTable, path } =
    staticAttachmentImportPaths[uploadSpec.staticPathKey];

  const validationQueryResource = generateInQueryResource(
    baseTable,
    'Batch Attachment Upload',
    [
      {
        field: { path },
        lookUp: uploadableFiles.map(({ uploadFile }) => uploadFile.parsedName),
      },
    ]
  );
  const rawValidationResponse = await validationPromiseGenerator(
    validationQueryResource
  );
  const mappedResponse = rawValidationResponse.map(([targetId, restResult]) => {
    const rawResult = uploadSpec.formatQueryResults(restResult[0]);
    return rawResult === undefined
      ? undefined
      : ([targetId, rawResult] as const);
  });

  return matchFileSpec(
    uploadableFiles,
    filterArray(mappedResponse),
    keepDisambiguation
  );
}

type MatchSelectedFiles = {
  readonly resolvedFiles: RA<PartialUploadableFileSpec>;
  readonly duplicateFiles: RA<PartialUploadableFileSpec>;
};
export const matchSelectedFiles = (
  previousUploadables: RA<PartialUploadableFileSpec>,
  filesToResolve: RA<PartialUploadableFileSpec>
): MatchSelectedFiles =>
  filesToResolve.reduce<MatchSelectedFiles>(
    (previousMatchedSpec, uploadable) => {
      const matchedIndex = previousMatchedSpec.resolvedFiles.findIndex(
        (previousUploadable) =>
          previousUploadable.uploadFile !== undefined &&
          previousUploadable.uploadFile.file.name ===
            uploadable.uploadFile.file.name &&
          previousUploadable.uploadFile.file.size ===
            uploadable.uploadFile.file.size &&
          previousUploadable.uploadFile.file.type ===
            uploadable.uploadFile.file.type
      );
      if (matchedIndex === -1)
        return {
          ...previousMatchedSpec,
          resolvedFiles: insertItem(
            previousMatchedSpec.resolvedFiles,
            previousMatchedSpec.resolvedFiles.length,
            uploadable
          ),
        };
      const previousMatch = previousMatchedSpec.resolvedFiles[matchedIndex];
      if (previousMatch.uploadFile.file instanceof File)
        return {
          ...previousMatchedSpec,
          duplicateFiles: [...previousMatchedSpec.duplicateFiles, uploadable],
        };

      return {
        ...previousMatchedSpec,
        resolvedFiles: replaceItem(
          previousMatchedSpec.resolvedFiles,
          matchedIndex,
          {
            ...previousMatch,
            uploadFile: uploadable.uploadFile,
            /*
             * Generating tokens again because the file could have been
             * uploaded to the asset server but not yet recorded in Specify DB.
             */
            uploadTokenSpec: undefined,
            /*
             * Take the new status in case of parse failure was reported.
             * But take the previous status it was a success
             */
            status:
              previousMatch.status?.type === 'success' ||
              (previousMatch.status?.type === 'skipped' &&
                previousMatch.status.reason === 'alreadyUploaded')
                ? previousMatch.status
                : uploadable.status,
          }
        ),
      };
    },
    {
      resolvedFiles: previousUploadables,
      duplicateFiles: [],
    }
  );

export function resolveFileNames(
  fileName: string,
  getFormatted: (rawName: number | string | undefined) => string | undefined,
  formatter?: UiFormatter
): string | undefined {
  // BUG: Won't catch if formatters begin or end with a space
  const splitName = stripFileExtension(fileName).trim();
  let nameToParse = splitName;
  if (
    formatter !== undefined &&
    formatter.fields.every(
      (field) => !(field instanceof formatterTypeMapper.regex)
    )
  ) {
    const formattedLength = formatter.fields.reduce(
      (length, field) => length + field.size,
      0
    );
    nameToParse = fileName.trim().slice(0, formattedLength);
  }
  let formatted = nameToParse === '' ? undefined : getFormatted(nameToParse);
  const numericFields = formatter?.fields.filter(
    (field) => field instanceof formatterTypeMapper.numeric
  );
  if (
    formatter?.fields?.length === 1 &&
    numericFields?.length === 1 &&
    formatted === undefined &&
    splitName !== ''
  ) {
    const numericValue = splitName.padStart(numericFields[0].size, '0');
    formatted = getFormatted(numericValue);
  }
  return formatted;
}

const validationPromiseGenerator = async (
  queryResource: SpecifyResource<SpQuery>
): Promise<RA<readonly [number, RA<number | string | null>]>> =>
  ajax<{
    // First value is the primary key
    readonly results: RA<readonly [number, ...RA<number | string | null>]>;
  }>('/stored_query/ephemeral/', {
    method: 'POST',
    headers: {
      // eslint-disable-next-line @typescript-eslint/naming-convention
      Accept: 'application/json',
    },
    body: keysToLowerCase({
      ...serializeResource(queryResource),
      countOnly: false,
      limit: 0,
    }),
  }).then(({ data }) =>
    data.results.map(([target, ...restResult]) => [target, restResult])
  );

export const matchFileSpec = (
  uploadFileSpec: RA<PartialUploadableFileSpec>,
  queryResults: RA<readonly [number, string]>,
  keepDisambiguation: boolean = false
): RA<PartialUploadableFileSpec> =>
  uploadFileSpec.map((spec) => {
    const specParsedName = spec.uploadFile?.parsedName;
    // Don't match files already uploaded.
    if (specParsedName === undefined || typeof spec.attachmentId === 'number')
      return spec;
    const matchingResults = queryResults.filter(
      (result) => result[1] === specParsedName
    );
    const newSpec: PartialUploadableFileSpec = {
      ...spec,
      matchedId: matchingResults.map((result) => result[0]),
    };

    if (
      keepDisambiguation &&
      spec.disambiguated !== undefined &&
      spec.matchedId !== undefined &&
      // If disambiguation was chosen, but it became invalid, reset disambiguation
      newSpec.matchedId?.includes(spec.disambiguated) === true &&
      spec.matchedId.length === newSpec.matchedId.length &&
      newSpec.matchedId.every((newMatch) => spec.matchedId!.includes(newMatch))
      // Reset disambiguation if sets are different in any way
    ) {
      return newSpec;
    }
    return { ...newSpec, disambiguated: undefined };
  });

const getBaseModelInField = (
  baseTable: keyof Tables,
  files: RA<PartialUploadableFileSpec>
) => ({
  field: {
    path: strictGetTable(baseTable).idField.name,
    isDisplay: false,
  },
  lookUp: filterArray(
    files.map((uploadable) =>
      uploadable.status?.type === 'matched' ? uploadable.status.id : undefined
    )
  ),
});

export async function reconstructDeletingAttachment(
  staticKey: keyof typeof staticAttachmentImportPaths,
  deletableFiles: RA<PartialUploadableFileSpec>
): Promise<RA<PartialUploadableFileSpec>> {
  const baseTable = staticAttachmentImportPaths[staticKey].baseTable;
  const relationshipName = `${baseTable}attachments`;
  const attachmentTableId = `${baseTable}attachmentid`;
  const path = `${relationshipName}.${attachmentTableId}`;
  const relatedAttachments = filterArray(
    deletableFiles.map((deletable) =>
      deletable.status?.type === 'matched' ? deletable.attachmentId : undefined
    )
  );
  const reconstructingQueryResource = generateInQueryResource(
    baseTable,
    'Batch Attachment Upload',
    [
      getBaseModelInField(baseTable, deletableFiles),
      {
        field: { path },
        lookUp: relatedAttachments,
      },
    ]
  );

  const queryResults = await validationPromiseGenerator(
    reconstructingQueryResource
  );
  return inferDeletedAttachments(queryResults, deletableFiles);
}

export async function reconstructUploadingAttachmentSpec(
  staticKey: keyof typeof staticAttachmentImportPaths,
  uploadableFiles: RA<PartialUploadableFileSpec>
): Promise<RA<PartialUploadableFileSpec>> {
  const baseTable = staticAttachmentImportPaths[staticKey].baseTable;
  const relationshipName = `${baseTable}attachments`;
  const pathToAttachmentLocation = `${relationshipName}.attachment.attachmentLocation`;
  const attachmentTableId = `${relationshipName}.${baseTable}attachmentid`;
  const filteredAttachmentLocations = filterArray(
    uploadableFiles.map((uploadable) =>
      uploadable.status?.type === 'matched'
        ? uploadable.uploadTokenSpec?.attachmentLocation
        : undefined
    )
  );

  const reconstructingQueryResource = generateInQueryResource(
    baseTable,
    'Batch Attachment Upload',
    [
      getBaseModelInField(baseTable, uploadableFiles),
      {
        field: {
          path: attachmentTableId,
          isDisplay: true,
          id: queryFieldFilters.any.id,
        },
      },
      {
        field: { path: pathToAttachmentLocation },
        lookUp: filteredAttachmentLocations,
      },
    ]
  );
  const queryResults = await validationPromiseGenerator(
    reconstructingQueryResource
  );
  return inferUploadedAttachments(queryResults, uploadableFiles);
}

export const inferUploadedAttachments = (
  queryResults: RA<readonly [number, RA<number | string | null>]>,
  uploadableFiles: RA<PartialUploadableFileSpec>
): RA<PartialUploadableFileSpec> =>
  uploadableFiles.map((uploadable) => {
    if (uploadable.status?.type !== 'matched') return uploadable;
    const matchedId = uploadable.status.id;
    const foundInQueryResult = queryResults.find(
      ([targetId, [_, attachmentLocation]]) =>
        typeof attachmentLocation === 'string' &&
        targetId === matchedId &&
        attachmentLocation.toString() ===
          uploadable.uploadTokenSpec!.attachmentLocation
    );

    return {
      ...uploadable,
      attachmentId: foundInQueryResult?.[1][0] as number,
      status:
        typeof foundInQueryResult === 'object'
          ? ({ type: 'success', successType: 'uploaded' } as const)
          : uploadable.status.type === 'matched'
          ? //
            /*
             *BUG: Handle case where attachment location is set to null or resource no longer exists better.
             * Currently, it will incorrectly inform it to be interrupted. That is fine since trying to upload
             * the dataset will automatically correctly regenerate tokens / show match error
             */
            ({
              type: 'cancelled',
              reason: 'uploadInterruption',
            } as const)
          : uploadable.status,
    };
  });

export const inferDeletedAttachments = (
  queryResults: RA<readonly [number, RA<number | string | null>]>,
  deletableFiles: RA<PartialUploadableFileSpec>
): RA<PartialUploadableFileSpec> =>
  deletableFiles.map((deletable) => {
    if (deletable.status?.type !== 'matched') return deletable;
    const matchedId = deletable.status.id;
    const foundInQueryResult = queryResults.find(
      ([targetId, [attachmentId]]) =>
        targetId === matchedId && attachmentId === deletable.attachmentId
    );
    return {
      ...deletable,
      attachmentId: foundInQueryResult?.[1][0] as number,
      status:
        foundInQueryResult === undefined
          ? ({ type: 'success', successType: 'deleted' } as const)
          : deletable.status.type === 'matched'
          ? ({
              type: 'cancelled',
              reason: 'rollbackInterruption',
            } as const)
          : deletable.status,
    };
  });

export const keyLocalizationMapAttachment = {
  incorrectFormatter: attachmentsText.incorrectFormatter(),
  noFile: attachmentsText.noFile(),
  uploaded: commonText.uploaded(),
  deleted: attachmentsText.deleted(),
  alreadyUploaded: attachmentsText.alreadyUploaded(),
  skipped: attachmentsText.skipped(),
  cancelled: attachmentsText.cancelled(),
  matchError: attachmentsText.matchError(),
  noAttachments: attachmentsText.noAttachments(),
  uploadInterruption: attachmentsText.frontEndInterruption({
    action: wbText.upload(),
  }),
  rollbackInterruption: attachmentsText.frontEndInterruption({
    action: wbText.rollback(),
  }),
  errorReadingFile: attachmentsText.errorReadingFile(),
  saveConflict: formsText.saveConflict(),
  unhandledFatalResourceError: attachmentsText.unhandledFatalResourceError(),
  nothingFound: formsText.nothingFound(),
  noMatch: attachmentsText.noMatch(),
  multipleMatches: attachmentsText.multipleMatches(),
  correctlyFormatted: attachmentsText.correctlyFormatted(),
  userStopped: attachmentsText.stoppedByUser(),
  interruptionStopped: attachmentsText.interruptionStopped(),
  errorFetchingRecord: attachmentsText.errorFetchingRecord(),
  saveError: attachmentsText.errorSavingRecord(),
  attachmentUploadError: attachmentsText.attachmentUploadError(),
} as const;

export function resolveAttachmentStatus(
  attachmentStatus: AttachmentStatus
): string {
  if ('reason' in attachmentStatus) {
    const reason = keyLocalizationMapAttachment[attachmentStatus.reason];
    return commonText.colonLine({
      label: keyLocalizationMapAttachment[attachmentStatus.type],
      value: reason,
    });
  }
  return attachmentStatus.type === 'success'
    ? keyLocalizationMapAttachment[attachmentStatus.successType]
    : '';
}

export const getAttachmentsFromResource = (
  baseResource: SerializedResource<Tables['CollectionObject']>,
  relationshipName: string
): {
  readonly key: keyof SerializedResource<Tables['CollectionObject']>;
  readonly values: RA<SerializedResource<Tables['CollectionObjectAttachment']>>;
} =>
  defined(
    mappedFind(Object.entries(baseResource), ([key, value]) =>
      key.toLowerCase() === relationshipName.toLowerCase()
        ? {
            key,
            values: value as RA<
              SerializedResource<Tables['CollectionObjectAttachment']>
            >,
          }
        : undefined
    )
  );

type RecordResponse =
  | State<
      'invalid',
      { readonly reason: keyof typeof keyLocalizationMapAttachment }
    >
  | State<'valid', { readonly record: SerializedResource<CollectionObject> }>;

const wrapAjaxRecordResponse = async (
  ajaxPromise: () => Promise<
    AjaxResponseObject<SerializedRecord<CollectionObject>>
  >,
  // Defines errors on which to not trigger a retry.
  statusMap: RR<number | 'fallback', keyof typeof keyLocalizationMapAttachment>,
  triggerRetry?: () => void
): Promise<RecordResponse> =>
  ajaxPromise().then(({ data, status }) => {
    if (statusMap[status] !== undefined)
      return { type: 'invalid', reason: statusMap[status] };
    if (status !== Http.OK) {
      triggerRetry?.();
      return { type: 'invalid', reason: statusMap.fallback };
    }
    return {
      type: 'valid',
      record: serializeResource(data),
    };
  });

const baseStatusMap = {
  [Http.NOT_FOUND]: 'nothingFound',
  [Http.CONFLICT]: 'saveConflict',
} as const;

export const fetchForAttachmentUpload = async (
  baseTableName: keyof Tables,
  matchId: number,
  triggerRetry?: () => void
) =>
  wrapAjaxRecordResponse(
    async () =>
      ajax<SerializedRecord<CollectionObject>>(
        `/api/specify/${baseTableName.toLowerCase()}/${matchId}/`,
        {
          headers: { Accept: 'application/json' },
          expectedErrors: Object.values(Http),
          errorMode: 'silent',
        }
      ),
    { ...baseStatusMap, fallback: 'errorFetchingRecord' },
    triggerRetry
  );

export const saveForAttachmentUpload = async (
  baseTableName: keyof Tables,
  matchId: number,
  data: Partial<SerializedResource<CollectionObject>>,
  triggerRetry?: () => void
) =>
  wrapAjaxRecordResponse(
    async () =>
      ajax<SerializedRecord<CollectionObject>>(
        `/api/specify/${baseTableName.toLowerCase()}/${matchId}/`,
        {
          method: 'PUT',
          body: keysToLowerCase(addMissingFields(baseTableName, data)),
          headers: { Accept: 'application/json' },
          errorMode: 'silent',
          expectedErrors: Object.values(Http),
        }
      ),
    { ...baseStatusMap, fallback: 'saveError' },
    triggerRetry
  );
