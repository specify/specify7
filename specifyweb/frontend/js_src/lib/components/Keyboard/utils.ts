import type { LocalizedString } from 'typesafe-i18n';

import type { RA } from '../../utils/types';
import { localized } from '../../utils/types';
import type { KeyboardShortcuts, ModifierKey } from './config';
import { keyboardModifierLocalization, keyboardPlatform } from './config';
import { keyJoinSymbol } from './context';

const localizedKeyJoinSymbol = ' + ';
export const localizeKeyboardShortcut = (shortcut: string): LocalizedString =>
  localized(
    shortcut
      .split(keyJoinSymbol)
      .map((key) => keyboardModifierLocalization[key as ModifierKey] ?? key)
      .join(localizedKeyJoinSymbol)
  );

/**
 * If there is a keyboard shortcut defined for current system, use it
 * (also, if current system explicitly has empty array of shortcuts, use it).
 *
 * Otherwise, use the keyboard shortcut from one of the other platforms if set,
 * but change meta to ctrl and vice versa as necessary.
 */
export function resolvePlatformShortcuts(
  shortcut: KeyboardShortcuts
): RA<string> | undefined {
  if ('platform' in shortcut) return shortcut[keyboardPlatform];
  else if ('other' in shortcut)
    return keyboardPlatform === 'windows'
      ? shortcut.other
      : shortcut.other?.map(replaceCtrlWithMeta);
  else if ('windows' in shortcut)
    return keyboardPlatform === 'other'
      ? shortcut.other
      : shortcut.other?.map(replaceCtrlWithMeta);
  else if ('mac' in shortcut) return shortcut.other?.map(replaceMetaWithCtrl);
  else return undefined;
}

const replaceCtrlWithMeta = (shortcut: string): string =>
  shortcut
    .split(keyJoinSymbol)
    .map((key) => (key === 'ctrl' ? 'meta' : key))
    .join(keyJoinSymbol);

const replaceMetaWithCtrl = (shortcut: string): string =>
  shortcut
    .split(keyJoinSymbol)
    .map((key) => (key === 'meta' ? 'ctrl' : key))
    .join(keyJoinSymbol);
