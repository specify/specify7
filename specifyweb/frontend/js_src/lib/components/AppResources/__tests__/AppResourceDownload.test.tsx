import { render } from '@testing-library/react';
import React from 'react';

import { mount } from '../../../tests/reactUtils';
import { f } from '../../../utils/functools';
import { LoadingContext } from '../../Core/Contexts';
import * as FilePickerModule from '../../Molecules/FilePicker';
import { AppResourceDownload } from '../EditorComponents';
import { testAppResources } from './testAppResources';

const mockDownload = jest.fn();

jest.spyOn(FilePickerModule, 'downloadFile').mockImplementation(mockDownload);

beforeEach(() => {
  mockDownload.mockClear();
});

describe('AppResourceDownload', () => {
  const resource = testAppResources.appResources[0];
  const testData = 'testString';

  test('simple render', () => {
    const { asFragment } = render(
      <LoadingContext.Provider value={f.void}>
        <AppResourceDownload data={testData} resource={resource} />
      </LoadingContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('download file', async () => {
    const { getByRole, user } = mount(
      <LoadingContext.Provider value={f.void}>
        <AppResourceDownload data={testData} resource={resource} />
      </LoadingContext.Provider>
    );
    const button = getByRole('button');
    await user.click(button);
    expect(mockDownload).toHaveBeenCalled();
    expect(mockDownload.mock.lastCall).toEqual([
      'preferences.properties',
      'testString',
    ]);
  });
});
