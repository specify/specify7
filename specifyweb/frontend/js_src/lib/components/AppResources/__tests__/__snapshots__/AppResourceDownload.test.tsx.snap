// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppResourceDownload simple render 1`] = `
<DocumentFragment>
  <button
    class="button rounded cursor-pointer active:brightness-80 px-4 py-2
    disabled:bg-gray-200 disabled:dark:ring-neutral-500 disabled:ring-gray-400 disabled:text-gray-500 
    dark:disabled:!bg-neutral-700 gap-2 inline-flex items-center capitalize justify-center shadow-sm dialog-icon-success hover:brightness-90 dark:hover:brightness-150 bg-[color:var(--success-button-color)] text-white whitespace-nowrap"
    type="button"
  >
    Download
  </button>
</DocumentFragment>
`;
